port/linux/omrvmem.c

```diff
@@ -1145,7 +1151,11 @@ default_pageSize_reserve_memory(struct OMRPortLibrary *portLibrary, void *addres
 			result = NULL;
 		} else {
 			/* Update identifier and commit memory if required, else return reserved memory */
-			update_vmemIdentifier(identifier, result, result, byteAmount, mode, pageSize, OMRPORT_VMEM_PAGE_FLAG_NOT_USED, OMRPORT_VMEM_RESERVE_USED_MMAP, category, fd);
+			uintptr_t allocator = OMRPORT_VMEM_RESERVE_USED_MMAP;
+			if (useBackingSharedFile) {
+				allocator = OMRPORT_VMEM_RESERVE_USED_MMAP_SHM;
+			}
+			update_vmemIdentifier(identifier, result, result, byteAmount, mode, pageSize, OMRPORT_VMEM_PAGE_FLAG_NOT_USED, allocator, category, fd);
```

Comment:
What about the double mapping at https://github.com/eclipse/omr/blob/82079d3d56995cfb3b2e330f35f4b3d7faa4a03b/port/linux/omrvmem.c#L1212, should it be using OMRPORT_VMEM_RESERVE_USED_MMAP_SHM as well?

Reply:
The contiguous representation of the arraylet is private but the arraylets themselves (with double mapping) are shared, so your suggestion is probably the right way to do it. I'll update it

Reply:
As I was saying in another comment, I did not think it's needed since would not normally try to decommit through this double mapped contiguous memory. 

But for some other potential users of this API that might try to decommit, I understand the concern. 

So, we can temporarily modify GC code and decommit memory (using contiguous address) just prior to unmapping it, and compare footprint (at the end of GC or end of a test) when allocator is as is, MMAP, hence decommited with DONTNEED advice vs MMAP_SHM, hence decommited through REMOVE advice.

Reply:
Some observations that I noticed while testing this fix. We don't actually need the commit part when double mapping an arraylet at:
https://github.com/eclipse/omr/blob/82079d3d56995cfb3b2e330f35f4b3d7faa4a03b/port/linux/omrvmem.c#L1214 
through line 1223 because that's already taken care by `get_protectionBits(mode)` at:
https://github.com/eclipse/omr/blob/82079d3d56995cfb3b2e330f35f4b3d7faa4a03b/port/linux/omrvmem.c#L1193
I tested and behaviour stayed the same if we remove the commit part.
And in regards to double mapped arraylets we do not decommit them anywhere, all we do is call the `vmem_free_memory()` API whenever they become garbage to free them, which doesn't decommit the memory. But the question still remains whether we should use `OMRPORT_VMEM_RESERVE_USED_MMAP_SHM` or `OMRPORT_VMEM_RESERVE_USED_MMAP`. The code as it stands wouldn't change in behaviour if either is used. Probably the best approach would be removing the commit from double mapping and update to `_MMAP_SHM` since the heap is using that same flag and double mapped arrays are shared through mmap?

Reply:
Yes, it seem more correct to use `_MMAP_SHM` even if the GC doesn't decommit that way. Something else could use the API and call decommit.

Reply:
About the commit, same reasoning. It's not only about what the GC needs. The APIs should operate correctly even if used for a different purpose. What if a user wanted to double map executable memory, then the commit is needed.

Reply:
On top of that I would probably add a call to `vmem_decommit_memory()` just before `vmem_free_memory()` whenever we are freeing the contiguous representation of a double mapped arraylet. In this case I would still choose the flag `OMRPORT_VMEM_RESERVE_USED_MMAP_SHM` because that forces `madvise()` to be called with `MADV_REMOVE` which tells the OS that we won't need that memory again (releases memory back to OS immediately). That's safe because the contiguous representation of the arraylet is created as an anonymous private memory through mmap (the OS chooses where to put) set by the flags here:
https://github.com/eclipse/omr/blob/82079d3d56995cfb3b2e330f35f4b3d7faa4a03b/port/linux/omrvmem.c#L1186
Hence, we do not know where the OS is going to put the contiguous representation of the next arraylet, so it's better to give the memory back to the OS so that the OS has that memory available for future usage (e.g. allocating more contiguous representation of arraylets)

Reply:
Correct, but looking at the API it seems the commit call here is redundant because we use `mmap` to create the contiguous representation of the arraylets with the same flags and all the commit call is doing is calling `mprotect()` to change the protection for the calling process's memory page, and the protection is already set by the `protectionFlags` here when we `mmap` it:
https://github.com/eclipse/omr/blob/82079d3d56995cfb3b2e330f35f4b3d7faa4a03b/port/linux/omrvmem.c#L1194
I'm probably missing something?

Reply:
ok, makes sense.

Reply:
I did ran some tests locally just to make sure the above is correct. I ran a few tests from my local C program where I simulate a heap with a few arraylet leaves in it. It looks like the madvise()/decommit call with `MADV_DONTNEED` is not sufficient to free the contiguous representation of the arraylet and `MADV_REMOVE` does free that memory:

Case 1:

1. Create arraylet leaves
2. Call double mapping (create contiguous representation of arraylet)
3. Call `madvise` with `MADV_DONTNEED` on the contiguous address
4. Check free physical memory <—— Memory NOT freed
5. Call `munmap` on the contiguous address
6. Check free physical memory <—— Memory NOT freed
Here physical memory is not freed with neither `munmap` or `madvise` call

Case 2:

1. Create arraylet leaves
2. Call double mapping (create contiguous representation of arraylet)
3. Call `madvise` with `MADV_REMOVE` on the contiguous address
4. Check free physical memory <—— Most memory freed
5. Call `munmap` on the contiguous address
6. Check free physical memory <—— All Memory freed

Most memory is freed in the `madvise` call but some remnants are freed after `munmap` call. This shows that `OMRPORT_VMEM_RESERVE_USED_MMAP_SHM` is the correct flag to use when dealing with double mapped arraylet lets.

Reply:
Thanks, Igor, I confirm I'm ok too with extra addition that Peter suggested.
@rwy0717 please proceed with further steps